File: leetcode/200-number-of-islands/main.go
package main

import (
	"fmt"
	"strconv"
)

// method 1: dfs, hashtable
// runtime: 32ms, beats 9.4%
func numIslands(grid [][]byte) int {
	visited := make(map[string]bool)
	result := 0
	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[0]); j++ {
			key := strconv.Itoa(i) + "," + strconv.Itoa(j)
			_, existed := visited[key]
			if grid[i][j] == '1' && !existed {
				dfs(grid, i, j, visited)
				result++
			}
		}
	}
	return result
}

func dfs(grid [][]byte, i int, j int, visited map[string]bool) {
	if i < 0 || i+1 > len(grid) || j < 0 || j+1 > len(grid[0]) {
		return
	}
	key := strconv.Itoa(i) + "," + strconv.Itoa(j)
	_, existed := visited[key]
	if grid[i][j] == '1' && !existed {
		key := strconv.Itoa(i) + "," + strconv.Itoa(j)
		visited[key] = true
		dfs(grid, i-1, j, visited)
		dfs(grid, i+1, j, visited)
		dfs(grid, i, j-1, visited)
		dfs(grid, i, j+1, visited)
	}
}

// method 2: dfs, dummy array
// runtime: 0ms, beats 100%
func numIslands1(grid [][]byte) int {
	visited := make([][]bool, len(grid))
	for i := range visited {
		visited[i] = make([]bool, len(grid[i]))
	}
	result := 0
	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[0]); j++ {
			if grid[i][j] == '1' && !visited[i][j] {
				dfs1(grid, i, j, visited)
				result++
			}
		}
	}
	return result
}

func dfs1(grid [][]byte, i int, j int, visited [][]bool) {
	if i < 0 || i+1 > len(grid) || j < 0 || j+1 > len(grid[0]) {
		return
	}
	if grid[i][j] == '1' && !visited[i][j] {
		visited[i][j] = true
		dfs1(grid, i-1, j, visited)
		dfs1(grid, i+1, j, visited)
		dfs1(grid, i, j-1, visited)
		dfs1(grid, i, j+1, visited)
	}
}

func main() {
	m := [][]byte{
		{'1', '1', '1', '1', '0'},
		{'1', '1', '0', '0', '0'},
		{'1', '1', '0', '1', '0'},
		{'0', '0', '0', '0', '1'},
	}
	fmt.Println(numIslands1(m))
}

File: leetcode/200-number-of-islands/main.py
"""
    1st approach: dfs, hashtable for visited island territories
    268ms beats 10.53%
"""


class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        res = 0
        seen = []
        for i in range(len(grid)):
            seen.append(len(grid[0])*[False])
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if seen[i][j] == False and grid[i][j] == '1':
                    self.dfs(grid, i, j, seen)
                    res += 1
        return res

    def dfs(self, grid, i, j, seen):
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or seen[i][j] == True:
            return
        seen[i][j] = True
        if grid[i][j] == '1':
            self.dfs(grid, i-1, j, seen)
            self.dfs(grid, i+1, j, seen)
            self.dfs(grid, i, j-1, seen)
            self.dfs(grid, i, j+1, seen)


"""
    follow-up: no recursion

    2nd approach: bfs

    Time    O(n)
    Space   O(n)
    264 ms, faster than 11.41%
"""


class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        res = 0
        seen = set()
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if (i, j) not in seen and grid[i][j] == '1':
                    self.bfs(grid, i, j, seen)
                    res += 1
        return res

    def bfs(self, grid, x, y, seen):
        q = []
        q.append((x, y))
        while len(q) > 0:
            i, j = q.pop(0)
            if i < 0 or i == len(grid) or j < 0 or j == len(grid[0]):
                continue
            if grid[i][j] == '0':
                continue
            if (i, j) in seen:
                continue
            seen.add((i, j))
            q.append((i-1, j))
            q.append((i+1, j))
            q.append((i, j-1))
            q.append((i, j+1))


"""
    follow-up: diagonal nodes
"""


class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        res = 0
        seen = []
        for i in range(len(grid)):
            seen.append(len(grid[0])*[False])
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if seen[i][j] == False and grid[i][j] == '1':
                    self.dfs(grid, i, j, seen)
                    res += 1
        return res

    def dfs(self, grid, i, j, seen):
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or seen[i][j] == True:
            return
        seen[i][j] = True
        if grid[i][j] == '1':
            self.dfs(grid, i-1, j, seen)
            self.dfs(grid, i+1, j, seen)
            self.dfs(grid, i, j-1, seen)
            self.dfs(grid, i, j+1, seen)
            # diagonal direction
            self.dfs(grid, i-1, j-1, seen)
            self.dfs(grid, i+1, j+1, seen)
            self.dfs(grid, i+1, j-1, seen)
            self.dfs(grid, i-1, j+1, seen)
